### PR TITLE
Port redis sharding to master

### PR DESCRIPTION
**Pull request by pcmoritz**

### CODE REVIEW

File: python/ray/services.py [hunk 1]
```diff
@@ -303,6 +304,7 @@ def start_redis(node_ip_address="127.0.0.1", port=None, num_retries=20,
   redis_client = redis.StrictRedis(host="127.0.0.1", port=port)
   # Wait for the Redis server to start.
   wait_for_redis_to_start("127.0.0.1", port)
+  redis_client.config_set("maxclients", "65536")
```
Comment (robertnishihara) [inline python/ray/services.py]: Maybe we should do this in a separate PR so we can get it merged right away.

Also, this kind of thing may throw an exception, so it should be inside of a `try` `except` block. Also, maybe we should try something like

```
try:
  for val in range(12, 17):
    redis_client.config_set("maxclients", str(2 ** val))
except:
  ...
```

so that we get the maximum number of clients that we can get.

File: python/ray/services.py [hunk 2]
```diff
@@ -808,6 +822,21 @@ def start_ray_processes(address_info=None,
...
+    for i in range(num_redis_shards):
+      redis_port, _ = start_redis(node_ip_address,
+                                  stdout_file=redis_stdout_file,
+                                  stderr_file=redis_stderr_file,
+                                  cleanup=cleanup)
+      redis_shards.append(address(node_ip_address, redis_port))
+    address_info["redis_shards"] = redis_shards
+    # Store redis shard information in the primary redis shard.
+    for shard in redis_shards:
+      redis_client.rpush("RedisShards", shard)
```
Comment (stephanie-wang) [inline python/ray/services.py]: Since we're storing redis shard information in the primary redis shard anyway, maybe we can avoid passing around the shard information to all the components, and have them look up the shard addresses on startup?

Comment (pcmoritz) [inline python/ray/services.py]: Yeah, that's a great idea! It will also make fault tolerance of the global control store easier in the future.

File: python/ray/experimental/state.py [hunk 3]
```diff
@@ -65,8 +66,25 @@ def _initialize_global_state(self, redis_ip_address, redis_port):
...
     self.redis_client = redis.StrictRedis(host=redis_ip_address,
                                           port=redis_port)
+    self.redis_clients = []
+    for ip_address_port in self.redis_client.lrange(
+            "RedisShards", start=0, end=-1):
+      shard_address, shard_port = ip_address_port.split(b":")
+      self.redis_clients.append(redis.StrictRedis(host=shard_address,
+                                                  port=shard_port))
+
+  def _execute_command(self, object_id, *args):
```
Comment (robertnishihara) [inline python/ray/experimental/state.py]: This is sometimes called with task IDs, so can we call the argument `identifier` or something like that?

Also, can we add a doc string like the following (with the appropriate line lengths)?

```python
"""Execute a Redis command on the appropriate Redis shard.

This command hashes the identifier to find the appropriate Redis shard
to query, and runs the command on that Redis server.

Args:
  identifier: The object ID or the task ID of that the query is about.
  args: The command to run.

Returns:
  The value returned by the redis command.
"""
```

File: python/ray/experimental/state.py [hunk 4]
```diff
@@ -65,8 +66,25 @@ def _initialize_global_state(self, redis_ip_address, redis_port):
...
+    client = self.redis_clients[object_id.redis_shard_hash() %
+                                len(self.redis_clients)]
+    return client.execute_command(*args)
+
+  def _keys(self, pattern):
```
Comment (robertnishihara) [inline python/ray/experimental/state.py]: Can we add the following docstring.

```python
"""Execute KEYS on all Redis shards and concatenate the results."""
```

File: python/ray/experimental/state.py [hunk 5]
```diff
@@ -110,22 +130,21 @@ def object_table(self, object_id=None):
...
+      object_info_keys = self._keys(OBJECT_INFO_PREFIX + "*")
+      object_location_keys = self._keys(OBJECT_LOCATION_PREFIX + "*")
       object_ids_binary = set(
           [key[len(OBJECT_INFO_PREFIX):] for key in object_info_keys] +
           [key[len(OBJECT_LOCATION_PREFIX):] for key in object_location_keys])
       results = {}
       for object_id_binary in object_ids_binary:
         results[binary_to_object_id(object_id_binary)] = self._object_table(
-            object_id_binary)
+            ray.local_scheduler.ObjectID(object_id_binary))
```
Comment (robertnishihara) [inline python/ray/experimental/state.py]: This is what the `binary_to_object_id` function does (which we use in the previous line)

File: python/ray/experimental/state.py [hunk 6]
```diff
@@ -184,14 +204,15 @@ def task_table(self, task_id=None):
...
+      return self._task_table(task_id)
     else:
-      task_table_keys = self.redis_client.keys(TASK_PREFIX + "*")
+      task_table_keys = self._keys(TASK_PREFIX + "*")
       results = {}
       for key in task_table_keys:
         task_id_binary = key[len(TASK_PREFIX):]
         results[binary_to_hex(task_id_binary)] = self._task_table(
-            task_id_binary)
+            ray.local_scheduler.ObjectID(task_id_binary))
```
Comment (robertnishihara) [inline python/ray/experimental/state.py]: should use `binary_to_object_id`

File: python/ray/services.py [hunk 7]
```diff
@@ -785,18 +790,18 @@ def start_ray_processes(address_info=None,
...
   redis_address = address_info.get("redis_address")
+  redis_shards = []
   if include_redis:
     redis_stdout_file, redis_stderr_file = new_log_files("redis",
                                                          redirect_output)
     if redis_address is None:
-      # Start a Redis server. The start_redis method will choose a random port.
+      # Start primary Redis. The start_redis method will choose a random port.
       redis_port, _ = start_redis(node_ip_address,
                                   stdout_file=redis_stdout_file,
```
Comment (robertnishihara) [inline python/ray/services.py]: This reuses the same log file for all redis shards. We probably want separate ones, right?

File: src/common/lib/python/common_extension.cc [hunk 8]
```diff
@@ -173,6 +173,11 @@ static PyObject *PyObjectID_richcompare(PyObjectID *self,
   return result;
 }
 
+static PyObject *PyObjectID_redis_shard_hash(PyObjectID *self) {
```
Comment (robertnishihara) [inline src/common/lib/python/common_extension.cc]: Can we document the fact that this hash function must be the same as the one used in `get_redis_context` in `redis.cc`?

File: src/common/state/redis.cc [hunk 9]
```diff
@@ -77,52 +78,77 @@ extern int usleep(useconds_t usec);
...
-                     const char *node_ip_address,
-                     int num_args,
-                     const char **args) {
-  /* Check that the number of args is even. These args will be passed to the
-   * RAY.CONNECT Redis command, which takes arguments in pairs. */
-  if (num_args % 2 != 0) {
-    LOG_FATAL("The number of extra args must be divisible by two.");
+redisAsyncContext *get_redis_context(DBHandle *db, UniqueID id) {
+  UniqueIDHasher index;
+  return db->contexts[index(id) % db->contexts.size()];
```
Comment (robertnishihara) [inline src/common/state/redis.cc]: We may want to document the fact that if we change this hash function, we also need to change `PyObjectID_redis_shard_hash` in `common_extension.cc`

File: src/common/net.cc [hunk 10]
```diff
@@ -11,3 +15,41 @@ int parse_ip_addr_port(const char *ip_addr_port, char *ip_addr, int *port) {
...
 }
+
+/* Return true if the ip address is valid. */
+bool valid_ip_address(const std::string &ip_address) {
+  struct sockaddr_in sa;
+  int result = inet_pton(AF_INET, ip_address.c_str(), &sa.sin_addr);
+  return result == 1;
+}
+
+bool parse_ip_addrs_ports(std::string ip_addrs_ports,
```
Comment (robertnishihara) [inline src/common/net.cc]: Do we use this anywhere?

**Review by robertnishihara — COMMENTED** [review]
I haven't looked at everything yet. Still reviewing.

**Comment by stephanie-wang** [comment]
Continued in #539.
